### PR TITLE
bpo-44151: In `linear_regression()`, rename `regressor` arg as `independent_variable`

### PR DESCRIPTION
**Pull request by zkneupper**
# [bpo-44151](https://bugs.python.org/issue44151): In `linear_regression()`, rename `regressor` arg as `independent_variable`

```
[bpo-44151](https://bugs.python.org/issue44151): In `linear_regression()`, rename `regressor` arg as `independent_variable`
```

This PR addresses one of the several suggestions in https://bugs.python.org/issue44151 .

https://bugs.python.org/issue44151

### CODE REVIEW

**Comment by pablogsal** [comment]
Hummm, this is technically a backwards incompatible change as people may be using the function with keyword parameters.

**Comment by zkneupper** [comment]
> Hummm, this is technically a backwards incompatible change as people may be using the function with keyword parameters.

Hi Pablo, my apologies. I thought people would just pass positional arguments to the function. I don't understand how people could use the function with keyword parameters.

**Comment by pablogsal** [comment]
> > Hummm, this is technically a backwards incompatible change as people may be using the function with keyword parameters.
> 
> Hi Pablo, my apologies. I thought people would just pass positional arguments to the function. I don't understand how people could use the function with keyword parameters.

Like any other function that allows keyword parameters:

```
linear_regression(regressor=..., dependent_variable=...)
```

**Comment by zkneupper** [comment]
> > > Hummm, this is technically a backwards incompatible change as people may be using the function with keyword parameters.
> > 
> > 
> > Hi Pablo, my apologies. I thought people would just pass positional arguments to the function. I don't understand how people could use the function with keyword parameters.
> 
> Like any other function that allows keyword parameters:
> 
> ```
> linear_regression(regressor=..., dependent_variable=...)
> ```

Thanks, I understand now.

Are there any good patterns that can be used to update a function signature while maintaining backwards compatibility?

**Comment by zkneupper** [comment]
> > > > Hummm, this is technically a backwards incompatible change as people may be using the function with keyword parameters.
> > > 
> > > 
> > > Hi Pablo, my apologies. I thought people would just pass positional arguments to the function. I don't understand how people could use the function with keyword parameters.
> > 
> > 
> > Like any other function that allows keyword parameters:
> > ```
> > linear_regression(regressor=..., dependent_variable=...)
> > ```
> 
> Thanks, I understand now.
> 
> Are there any good patterns that can be used to update a function signature while maintaining backwards compatibility?

Could we do something like this within the standard library to ensure backward compatibility when an argument name is updated in a function definition:

https://gist.github.com/rfezzani/002181c8667ec4c671421a4d938167eb

**Comment by zkneupper** [comment]
> Hummm, this is technically a backwards incompatible change as people may be using the function with keyword parameters.

The current function signature `def linear_regression(regressor, dependent_variable, /):` has a forward slash `/` that makes `regressor` and `deppendent_variable` "positional only parameters" (https://www.python.org/dev/peps/pep-0436/#functions-with-positional-only-parameters).

Doesn't that mean that users can't use the function with keyword parameters already?

**Comment by rhettinger** [comment]
Please hold off on the PR until the tracker discussion is complete.  We may yet switch to *x* and *y*.  And other improvements may be made as well.

There are no backwards compatibility issues because 1) linear_regression() is new hasn't been release yet and 2) the parameters are positional-only so they can't be called with keyword arguments.

**Comment by rhettinger** [comment]
Zachery, unless someone steps with an objection, I think you can go forward with this signature:

     linear_regression(x, y, /) -> LinearRegression(slope, intercept)

File: Doc/library/statistics.rst [hunk 1]
```diff
@@ -643,23 +643,23 @@ However, for reading convenience, most of the examples show sorted sequences.
...
    .. versionadded:: 3.10
 
-.. function:: linear_regression(regressor, dependent_variable)
+.. function:: linear_regression(independent_variable, dependent_variable)
 
    Return the intercept and slope of `simple linear regression
    <https://en.wikipedia.org/wiki/Simple_linear_regression>`_
    parameters estimated using ordinary least squares. Simple linear
-   regression describes the relationship between *regressor* and
+   regression describes the relationship between *independent_variable* and
```
Comment (rhettinger) [inline Doc/library/statistics.rst]: Add articles to the sentence:  between AN independent variable and A dependent variable.

Also the example on line 675 needs to be updated to reflect the new slope/intercept ordering.

File: Doc/library/statistics.rst [hunk 2]
```diff
@@ -643,23 +643,23 @@ However, for reading convenience, most of the examples show sorted sequences.
...
+      *dependent_variable = intercept + slope \* independent_variable + noise*
 
    where ``intercept`` and ``slope`` are the regression parameters that are
    estimated, and noise represents the
    variability of the data that was not explained by the linear regression
    (it is equal to the difference between predicted and actual values
    of dependent variable).
 
-   Both inputs must be of the same length (no less than two), and regressor
+   Both inputs must be of the same length (no less than two), and independent_variable
```
Comment (rhettinger) [inline Doc/library/statistics.rst]: AN independent variable.

File: Lib/statistics.py [hunk 3]
```diff
@@ -932,18 +932,18 @@ def correlation(x, y, /):
         raise StatisticsError('at least one of the inputs is constant')
 
 
-LinearRegression = namedtuple('LinearRegression', ['intercept', 'slope'])
+LinearRegression = namedtuple('LinearRegression', ['slope', 'intercept'])
```
Comment (rhettinger) [inline Lib/statistics.py]: Minor nit, make this a tuple instead of a list:  ``('slope', 'intercept')``.

File: Lib/statistics.py [hunk 4]
```diff
@@ -932,18 +932,18 @@ def correlation(x, y, /):
...
 
 
-def linear_regression(regressor, dependent_variable, /):
+def linear_regression(x, y, /):
     """Intercept and slope for simple linear regression
 
     Return the intercept and slope of simple linear regression
     parameters estimated using ordinary least squares. Simple linear
-    regression describes relationship between *regressor* and
+    regression describes relationship between *x* and
```
Comment (rhettinger) [inline Lib/statistics.py:943]: On the next line, change "dependent variable" to "y".

**Review by rhettinger — CHANGES_REQUESTED** [review]

File: Lib/statistics.py (hunk 3 above)
Comment (zkneupper) [reply Lib/statistics.py]: Done.

File: Lib/statistics.py (hunk 4 above)
Comment (zkneupper) [reply Lib/statistics.py:943]: Done.

File: Doc/library/statistics.rst (hunk 1 above)
Comment (zkneupper) [reply Doc/library/statistics.rst]: I've updated the example on line 675 (now line 676).

I made some additional changes to 651:
... regression describes the relationship between an independent variable *x* and a dependent variable *y* ...

**Comment by zkneupper** [comment]
I have made the requested changes; please review again.

**Pull request merged**